smartconfig/entry.py

```diff
@@ -0,0 +1,123 @@
+        Add the `__defined_attributes` and `__path_override` attributes to the new entry.
+
+        Args:
+            path: Custom path used for this entry instead of the module name.
+        """
+        dict_[f"{cls.__name__}__path_override"] = path
+
+        return super().__new__(cls, name, bases, dict_)
+
+    def _register_entry(cls) -> None:
+        """Set the `__path` attribute and register the entry."""
+        cls.__path = cls.__path_override or cls.__module__
+        if cls.__path in _registry.used_paths:
+            raise PathConflict(f"An entry at {cls.__path!r} already exists.")  # TODO: Add an FAQ link.
+
+        configuration = {
+            key: value for key, value in cls.__dict__.items() if not key.startswith('_')
+        }
+
+        current_node = _registry.global_configuration
+
+        for node_name in cls.__path.split("."):
+            current_node = current_node.setdefault(node_name, {})
+
+        for key, value in configuration.items():
+            # We only write values that aren't already defined.
+            if key not in current_node:
+                current_node[key] = value
+
+        used_paths.add(cls.__path)
+
+    def _check_undefined_entries(cls) -> None:
+        """Raise `ConfigurationKeyError` if any attribute doesn't have a defined value."""
+        defined_attributes = [
+            name for name in chain(cls.__dict__.keys(), getattr(cls, "__annotations__", ())) if not name.startswith('_')
+        ]
+
+        for attribute in defined_attributes:
+            try:
+                get_attribute(cls.__path, attribute)
+            except (ConfigurationError, ConfigurationKeyError):
+                raise ConfigurationKeyError(f"Attribute {attribute!r} doesn't have a defined value.") from None
+
+    def __init__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any], path: Optional[str] = None):
+        """
+        Initialize the new entry.
+
+        Raises:
+            PathConflict: An entry is already registered for this path. Use the `path` metaclass argument.
+            ConfigurationKeyError: An attribute doesn't have a defined value.
+        """
+        super().__init__(name, bases, dict_)
+
+        cls._register_entry()
+        cls._check_undefined_entries()
+
+    def __del__(cls) -> None:
+        """Cleanup the defaults from the global configuration."""
+        unload_defaults(cls.__path)
+
+    def __repr__(cls) -> str:
+        """Return a short representation of the entry."""
+        if hasattr(cls, '__path'):
+            return f"<_ConfigEntryMeta {cls.__name__} at {cls.__path!r}>"
+        else:
+            return f"<_ConfigEntryMeta {cls.__name__}>"
+
+    def __eq__(cls, other: Any) -> bool:
+        """Return true if this entry and the other point to the same path."""
+        if not isinstance(other, _ConfigEntryMeta):
+            return NotImplemented
+        return cls.__path == other.__path
+
+
+class ConfigEntry(metaclass=_ConfigEntryMeta):
```

Comment:
The entry will be registered at the path `smartconfig`, it shouldn't really be an error since there is nothing to override there.

Reply:
Should we have a special check to avoid registering that path then? I don't know of a better way to avoid it.

Reply:
Do we even want to avoid it? I don't find it problematic, it will just hold an empty dictionary in the global configuration.

Reply:
I'm not sure. It may be strange to see if something external needs access to the config, like for introspection. I suppose we can remove it later if it becomes a problem.